### PR TITLE
Do not allow to reuse TOTP passcode

### PR DESCRIPTION
**Pull request by lafriks**

### CODE REVIEW

**Comment by ghost** [comment]
Thank you for taking care of the security issues reported, even those with lower priority like this one.

**Comment by techknowlogick** [comment]
~LGTM~

edit: noticed a minor typo

File: models/migrations/migrations.go [hunk 1]
```diff
@@ -176,6 +176,8 @@ var migrations = []Migration{
 	NewMigration("add is_fsck_enabled column for repos", addFsckEnabledToRepo),
 	// v61 -> v62
 	NewMigration("add size column for attachments", addSizeToAttachment),
+	// v62 -> v63
+	NewMigration("add last used passcode column for TOPT", addLastUsedPasscodeTOPT),
```
Comment (techknowlogick) [inline models/migrations/migrations.go]: Incorrect acronym, this should be `TOTP` (per: https://en.wikipedia.org/wiki/Time-based_One-time_Password_algorithm). There are other places in this PR that have the incorrect acronym.

Sorry for noticing this after my LG-TM, this should be an easy fix though.

**Review by techknowlogick — CHANGES_REQUESTED** [review]

**Review by jonasfranz — APPROVED** [review]
LGTM except the typo reported

**Comment by lafriks** [comment]
@techknowlogick fixed

**Review by appleboy — APPROVED** [review]

**Review by techknowlogick — APPROVED** [review]

**Comment by techknowlogick** [comment]
thanks @lafriks, I've just approved.

**Review by daviian — CHANGES_REQUESTED** [review]
Can we just store the hash of it instead of the plain passcode? Seems to be a more trustable approach to solve this.
My gut says this should be handled as sensitive information.

**Comment by lafriks** [comment]
@daviian it is one time usage passcode. Hashing would not give much anyway as it could be easily bruteforced back to normal value (only six digits so that would take few seconds to get value back from hash), I don't think it would change much if anything

**Review by daviian — APPROVED** [review]

**Comment by daviian** [comment]
Don't want to block that PR ;)

But if it's that easy to bruteforce TOTP, why use it in the first place?

**Comment by techknowlogick** [comment]
@daviian I think @lafriks means that it is easy to bruteforce a hash of 6 characters, so if an attacker has access to the DB a hash wouldn't really stop them from knowing what the token was. This is fine because tokens change every 30 seconds, and this is a protection just in case someone MITM a user to intercept the token so it can't be re-used.

**Comment by daviian** [comment]
@techknowlogick @lafriks You are totally right! :+1:

**Comment by lafriks** [comment]
As this is more enchantment and it is hard to backport to 1.4 without breaking upgrade process, I will not backport it

**Comment by ghost** [comment]
Personally I don't care too much, but may I ask why finders don't receive any credits?  Indeed, it would be really appreciated...

**Comment by techknowlogick** [comment]
@cezar97 I can't speak on behalf of the other maintainers, but I've asked for a thanks to be added to the blog post for the release notes. As only owners see the security reports I wasn't able to add the appropriate thanks to my PR for the release blog post.

**Comment by lafriks** [comment]
@cezar97 finders will get credits in blog post

**Pull request merged**